polyply/src/gen_seq.py

```diff
@@ -0,0 +1,254 @@
+# Copyright 2020 University of Groningen
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import random
+import json
+import networkx as nx
+from networkx.readwrite import json_graph
+from .load_library import load_library
+from .generate_templates import find_atoms
+
+
+def _branched_graph(resname, branching_f, n_levels):
+    """
+    Generate a tree graph of branching factor `branching_f`
+    and number of generations as `n_levels` - 1. Each node
+    gets the attribute resname.
+
+    Parameters
+    ----------
+    resname:  str
+    branching_f: int
+    n_levels: int
+
+    Returns
+    -------
+    `:class:networkx.Graph`
+    """
+    graph = nx.balanced_tree(r=branching_f, h=n_levels-1)
+    resnames = {idx: resname for idx in graph.nodes}
+    nx.set_node_attributes(graph, resnames, "resname")
+    return graph
+
+
+def _random_replace_nodes(graph, residues, weights):
+    """
+    Randomly replace resname attribute of `graph`
+    based on the names in `residues` taking into
+    account `weights` provided.
+
+    Parameters
+    ----------
+    graph: `:class:networkx.Graph`
+    residues: list[str]
+    weights: list[float]
+
+    Returns:
+    --------
+    `:class:networkx.Graph`
+    """
+    for node in graph.nodes:
+        resname = random.choices(residues, weights=weights)
+        graph.nodes[node]["resname"] = resname[0]
+
+    return graph
+
+
+class MacroString():
+    """
+    Define a (random) tree graph based on a string.
+    The graph is generated each time the `gen_graph`
+    attribute is run. For random graph they will be
+    independent.
+    """
+
+    def __init__(self, string):
+        """
+        Convert string into definitions of the
+        random tree graph.
+
+        Parameters:
+        -----------
+        string: str
+           the input string containing all definitions
+        """
+        input_params = string.split(":")
+        self.name = input_params[0]
+        self.levels = int(input_params[1])
+        self.bfact = int(input_params[2])
+        self.residues = []
+        self.weights = []
+
+        for res_prob in input_params[3].split(','):
```

Comment:
First disect input_params into useful bits, and all in the same place.

Reply:
that's what this function is doing ...

Reply:
Wrong order. First disect, all in one place, *then* set self.residues and self.weights, **then** the for loop.